webauthn-rp-proxy/src/main.rs

```diff
@@ -0,0 +1,202 @@
+use clap::{App, Arg};
+use serde::{Deserialize, Serialize};
+use serde_json::Result;
+use std::io::{self, Read};
+use uuid::Uuid;
+use webauthn_rs::prelude::*;
+
+// A simple command-line wrapper around webauthn-rs
+
+// There are four commands, each representing a server-side step in
+// the basic webauthn protocol:
+
+// - register-start
+// - register-finish
+// - authenticate-start
+// - authenticate-finish
+
+// The program's --step argument specifies which step from the list
+// above is being invoked.  It reads the corresponding JSON Request
+// struct below from stdin, carries out the webauthn step, and writes
+// the JSON Response struct to stdout if no errors occur.
+
+// The Response structs contain fields labeled client and server.  The
+// value in the client field should be sent to the browser.  The value
+// in the server field should be used on the server, AND SHOULD NOT BE
+// SENT TO THE CLIENT LEST SECURITY BE COMPROMISED.
+
+// The purpose of this program is twofold: to facilitate testing, and
+// to make it easy to implement webauthn in other programming
+// languages by doing simple JSON I/O to this program.
+
+#[derive(Deserialize)]
+struct RegisterStartRequest {
+    exclude_credentials: Vec<CredentialID>,
+    rp_id: String,
+    rp_origin: String,
+    user_display_name: String,
+    user_name: String,
+    uuid: String,
+}
+
+#[derive(Serialize)]
+struct RegisterStartResponse {
+    client: CreationChallengeResponse,
+    server: PasskeyRegistration,
+}
+
+#[derive(Deserialize)]
+struct RegisterFinishRequest {
+    register_public_key_credential: RegisterPublicKeyCredential,
+    passkey_registration: PasskeyRegistration,
+    rp_id: String,
+    rp_origin: String,
+}
+
+#[derive(Serialize)]
+struct RegisterFinishResponse {
+    server: Passkey,
+}
+
+#[derive(Deserialize)]
+struct AuthenticateStartRequest {
+    passkeys: Vec<Passkey>,
+    rp_id: String,
+    rp_origin: String,
+}
+
+#[derive(Serialize)]
+struct AuthenticateStartResponse {
+    client: RequestChallengeResponse,
+    server: PasskeyAuthentication,
+}
+
+#[derive(Deserialize)]
+struct AuthenticateFinishRequest {
+    passkey_authentication: PasskeyAuthentication,
+    public_key_credential: PublicKeyCredential,
+    rp_id: String,
+    rp_origin: String,
+}
+
+#[derive(Serialize)]
+struct AuthenticateFinishResponse {
+    server: AuthenticationResult,
+}
+
+fn main() {
+    let arguments = App::new("webauthn-rp-proxy")
+	.arg(
+	    Arg::with_name("step")
+		.long("--step")
+		.takes_value(true)
+		.required(true)
+		.possible_values(&["authenticate-start",
```

Comment:
I would use `clap`'s derive macro with `Subcommand` here, eg:

```rust
#[derive(Debug, Subcommand)]
pub enum Step {
    AuthenticateStart,
    AuthenticateEnd,
    // TODO
}

#[derive(Debug, clap::Parser)]
#[clap(about = "TODO")]
pub struct CliParser {
    #[clap(subcommand)]
    pub step: Step,
    // add other global args here, eg:
    #[clap(long)]
    pub pretty_print: bool,
}
```

Then using it becomes much simpler:

```rust
let opt = CliParser::parse();
let r = match opt.step {
    Step::AuthenticateStart => authenticate_start(&buffer).unwrap();
    // add other steps
};
```

Everything is checked at compile time, so no need to have string constants or `unreachable!()`.

You should also check the result from those functions (ie: `unwrap()` or `expect()`), because they're passing back errors from serde's serialisation library, and then you've ignored them (with `let _ = ...`). While that silences clippy, but doesn't actually fix the issue. 😄

Reply:
Thank you very much for the thorough review.  I'm a novice at Rust, and have learned a lot from working through your feedback.

I ended up rebasing and pushing again, and Github no longer displays your review comments in the Files changed view.  Sorry about that.  But at least they're still here, in the Conversation page.

I've switched to Subcommand, as suggested, and have captured errors as JSON, too.

Reply:
Thanks for sticking with this. 😄